### ISSUE TITLE
need a server side implmentation in golang for realtime prediction(ad ctr)

### ISSUE THREAD

**Issue by dwroy** [dmlc/xgboost#3107]
I am using spark train models

plan to use golang for prediction

need a golang implmentation as a predictor server

 any ideas?  or any api doc for  xgboost.so lib?

**Comment by hcho3** [comment]
Take a look at the header [c_api.h](https://github.com/dmlc/xgboost/blob/master/include/xgboost/c_api.h), which list all functions exposed by xgboost.so library. In particular, the functions `XGBoosterCreate`, `XGBoosterPredict` and `XGDMatrixCreateFromFile`. The C API functions are not well documented, unfortunately.

Alternatively, you may consider using [dmlc/treelite](https://github.com/dmlc/treelite). This library is dedicated to model deployment. For now, treelite requires that you compile your model into a shared library. The option of deploying your model without compilation is currently on the docket; see issue [#11](https://github.com/dmlc/treelite/issues/11).

**Comment by dwroy** [comment]
thank you so much. And  I wonder after training model in spark, how to access those params/weights for each DTree, and the weights of the logistic model (I use reg:logsitic). I can hardly find some docs. maybe I could try a lite version in golang

**Comment by dwroy** [comment]
need some help here, I use gcc-7 compile xgboost.a, and link in my project, when I run make my project, I got this

**Comment by dwroy** [comment]
need some help here, I use gcc-7 compile xgboost.a, and link in my project, when I run make my project, I got this  `Undefined symbols for architecture x86_64:
  "_GOMP_atomic_end", referenced from:
      xgboost::metric::EvalMClassBase<xgboost::metric::EvalMatchError>::Eval(std::vector<float, std::allocator<float> > const&, xgboost::MetaInfo const&, bool) const [clone ._omp_fn.0] in libxgboost.a(multiclass_metric.o)
      xgboost::metric::EvalMClassBase<xgboost::metric::EvalMultiLogLoss>::Eval(std::vector<float, std::allocator<float> > const&, xgboost::MetaInfo const&, bool) const [clone ._omp_fn.1] in libxgboost.a(multiclass_metric.o)
      xgboost::metric::EvalEWiseBase<xgboost::metric::EvalRMSE>::Eval(std::vector<float, std::allocator<float> > const&, xgboost::MetaInfo const&, bool) const [clone ._omp_fn.0] in libxgboost.a(elementwise_metric.o)
      xgboost::metric::EvalEWiseBase<xgboost::metric::EvalMAE>::Eval(std::vector<float, std::allocator<float> > const&, xgboost::MetaInfo const&, bool) const [clone ._omp_fn.1] in libxgboost.a(elementwise_metric.o)
      xgboost::metric::EvalEWiseBase<xgboost::metric::EvalError>::Eval(std::vector<float, std::allocator<float> > const&, xgboost::MetaInfo const&, bool) const [clone ._omp_fn.3] in libxgboost.a(elementwise_metric.o)
      xgboost::metric::EvalEWiseBase<xgboost::metric::EvalGammaDeviance>::Eval(std::vector<float, std::allocator<float> > const&, xgboost::MetaInfo const&, bool) const [clone ._omp_fn.5] in libxgboost.a(elementwise_metric.o)
      xgboost::metric::EvalEWiseBase<xgboost::metric::EvalLogLoss>::Eval(std::vector<float, std::allocator<float> > const&, xgboost::MetaInfo const&, bool) const [clone ._omp_fn.2] in libxgboost.a(elementwise_metric.o)
      ...`

**Comment by hcho3** [comment]
Did you link -lgomp to your project?

**Comment by hcho3** [comment]
Add the following to CMakeLists.txt:
```cpp
find_package(OpenMP)
if (OPENMP_FOUND)
    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
    set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}")
endif()
```

**Comment by dwroy** [comment]
**Last for my osx  I brew install  llvm  use its clang-5 as compiler( with openmp), CMakeList like below:**
`
set(CMAKE_C_COMPILER "/usr/local/Cellar/llvm/5.0.1/bin/clang")
set(CMAKE_CXX_COMPILER "/usr/local/Cellar/llvm/5.0.1/bin/clang++")
set(OPENMP_LIBRARIES "/usr/local/Cellar/llvm/5.0.1/lib")
set(OPENMP_INCLUDES "/usr/local/Cellar/llvm/5.0.1/include")

OPTION (USE_OpenMP "Use OpenMP to enamble <omp.h>" ON)
# Find OpenMP
if(APPLE AND USE_OpenMP)
    if(CMAKE_C_COMPILER_ID MATCHES "Clang")
        set(OpenMP_C "${CMAKE_C_COMPILER}")
        set(OpenMP_C_FLAGS "-fopenmp=libomp -Wno-unused-command-line-argument")
        set(OpenMP_C_LIB_NAMES "libomp" "libgomp" "libiomp5")
        set(OpenMP_libomp_LIBRARY ${OpenMP_C_LIB_NAMES})
        set(OpenMP_libgomp_LIBRARY ${OpenMP_C_LIB_NAMES})
        set(OpenMP_libiomp5_LIBRARY ${OpenMP_C_LIB_NAMES})
    endif()
    if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
        set(OpenMP_CXX "${CMAKE_CXX_COMPILER}")
        set(OpenMP_CXX_FLAGS "-fopenmp=libomp -Wno-unused-command-line-argument")
        set(OpenMP_CXX_LIB_NAMES "libomp" "libgomp" "libiomp5")
        set(OpenMP_libomp_LIBRARY ${OpenMP_CXX_LIB_NAMES})
        set(OpenMP_libgomp_LIBRARY ${OpenMP_CXX_LIB_NAMES})
        set(OpenMP_libiomp5_LIBRARY ${OpenMP_CXX_LIB_NAMES})
    endif()
endif()

if(USE_OpenMP)
    find_package(OpenMP REQUIRED)
endif(USE_OpenMP)

if (OPENMP_FOUND)
    include_directories("${OPENMP_INCLUDES}")
    link_directories("${OPENMP_LIBRARIES}")
    set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
    set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
    # set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}")
endif(OPENMP_FOUND)`

**but now come across another error, would you please take o look and help me? I am going crazy:** 

`ld: warning: could not create compact unwind for _XGDMatrixCreateFromCSR: dwarf uses DW_CFA_GNU_args_size
ld: warning: could not create compact unwind for _XGDMatrixCreateFromCSC: dwarf uses DW_CFA_GNU_args_size
Undefined symbols for architecture x86_64:
  "_RabitLinkTag", referenced from:
      __GLOBAL__sub_I_c_api.cc in libxgboost.a(c_api.o)
  "__gnu_parallel::_Settings::get()", referenced from:
      void std::__parallel::sort<__gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >, xgboost::MetaInfo::LabelAbsSort() const::'lambda'(unsigned long, unsigned long), __gnu_parallel::default_parallel_tag>(__gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >, __gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >, xgboost::MetaInfo::LabelAbsSort() const::'lambda'(unsigned long, unsigned long), __gnu_parallel::default_parallel_tag) in libxgboost.a(regression_obj.o)
      xgboost::metric::EvalAuc::Eval(std::vector<float, std::allocator<float> > const&, xgboost::MetaInfo const&, bool) const in libxgboost.a(rank_metric.o)`

**Comment by dwroy** [comment]
I give up my macbook,  build code in centos and sailed through， I will deploy my project in linux any way.   but how to work on mac is my pain

**Comment by bububa** [comment]
Golang c API wrapper  https://github.com/bububa/xgboost-go

**Comment by dmitryikh** [comment]
Pure Golang implementation for trees ensemble predictions for XGBoost & LightGBM:
https://github.com/dmitryikh/leaves
